### PR TITLE
TB NLnet - project theoretical basics

### PR DESCRIPTION
**Pull request by piotr-kleins**

### CODE REVIEW

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 1]
```diff
@@ -0,0 +1,419 @@
+---
+title: Open Source DRTM with TrenchBoot. Project's basics.
+abstract: Today, we will get you into project structure and environment. You
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: We don't know when someone will read this post. I would suggest to change abstract to something like this:

```
This is the first blog post from the Trenchboot series. It will introduce you to the project, its
structure and environment. Additionally the reader will find out more about each project component,
how to setup the environment and configure the build. 
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+          will find out more about each component and how properly configure
+          your build.
+cover: /covers/
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Do not forget to provide a cover image

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+categories:
+  - Firmware
+  - Security
+
+---
+
+If you haven't read [introductory blog post](link-to-introductory) yet, I
+strongly recommend to do it before proceeding here. We have presented our
+motivation, project's overview concept and goals there. Generally speaking,
+over time, we will be introducing new issues according to project's state. If
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
we will be introducing new issues according to project's state over time. If
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+you want to keep up with project development, stay tuned and always try to catch
+up missed articles.
+
+> Those articles will not always be strictly technical. Therefore, if you think
+some topics should be elaborated, feel free to ask us a question.
+
+In this article, I will explain how the project is build. Mostly, what hardware,
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
In this article, I will explain how the project is built. Mostly, what hardware,
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+firmware and software components we are using. The most essential parts of
+system will be also described in details. It's important to understand their
+role, so you will be able to, fully consciously, reproduce the environment.
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
role, so you will be able to reproduce the environment smoothly and consciously.
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+
+## Testing platform
+
+As you already know, our solution is built for AMD processors. Therefore, for
+tests we use suitable platforms. Our hardware infrastructure is still being
+expanded. Eventually, there will be few different devices with different
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
expanded. Eventually, there will be few more devices with different
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+characteristic. For now, we use **PC Engines apu2** which has **AMD GX-412TC**
+processor. We use connection via serial port, so every relevant output will
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
processor. We use connection via serial port, so every relevant output will be
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+presented in form of logs.
+
+## Operating system
+
+To better understand next section, first I will introduce operating system which
+we use. Our choice is [NixOS](https://nixos.org/nixos/) which is a Linux
+distribution based on unusual (and reliable) package management. You can install
+tools from official packages or build your own ones. Second feature will be
+strongly utilizes in entire project. Later, I will show you how to build custom
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
strongly utilized in entire project. Later, I will show you how to build custom
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+package on `flashrom` example.
+
+> Of course, you can use any other Linux distribution. However, every procedure
+will be carried out by us in NixOS. If you won't have all dedicated tools or
+wrong configuration, we can't provide full reliability.
+
+#### NixOS installation
+
+Now, I will guide through installation of NixOS. After it, you should have
+operating system in default configuration. Later, I will show you how to
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
an operating system in a default configuration. Later, I will show you how to
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+
+3. Move to the `flashrom2` directory and create a `default.nix` config file.
+
+```bash
+cd pkgs/tools/misc/flashrom2
+touch default.nix
+vim default.nix
+```
+
+4. Fill `default.nix` file with following content
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: I would describe what these mean. Probably these are dependencies, right?

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+our platform (both firmware and operating system). Before, we do that, I need to
+introduce some concepts. The better you understand them, the more awareness
+about security of your platform you will have and better understand the purpose
+of entire project. Let's break our entire system into following main parts. They
+are listed in order of execution when the platform is booted, so you can also
+see the flow and relations between them.
+
+#### BIOS
+
+Our solution is fully open-source, so we also use such firmware (BIOS) -
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
Our solution is open-source, so we also use such firmware (BIOS) -
```
We still have AGESA blob, unfortunately.

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+**coreboot**. Also, as we are maintainers of coreboot for PC Engines platforms,
+we use
+[this](https://github.com/pcengines/coreboot/tree/pcengines_trenchboot_4.11.x)
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Since we launch GRUB from disk, the branch shouldn't matter, it could be a typical release like v4.11.0.4 for example.

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+#### GRUB
+
+When coreboot ends its work, a bootloader comes in. We use **GRUB**. Its task is
+to boot operating system. Normally, if NixOS is installed on hard disk, GRUB
+loads it to RAM. Being more precise, *Linux kernel* and *initrd* are being
+loaded. However, that operation is slightly different when we enable DRTM. As it
+is main goal of our project, let's take a look a little bit closer on it.
+
+We prepared extension module called `slaunch`. When DRTM is enabled, `slaunch`
+executes AMD's dedicated machine instruction `SKINIT`. Treat is 'pre-DRTM'
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
executes AMD's dedicated machine instruction `SKINIT`. Treat it as 'pre-DRTM'
```
This is what you meant?

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+operation. It prepares all components and jump to secure loader - `Landing
+Zone`.
+
+#### Landing Zone (LZ)
+
+LZ has 2 main tasks: measure Linux kernel and then run it. If you wonder, when
+measurements are done, then you have just find out. Without LZ, DRTM doesn't
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
measurements are done, then you have just found out. Without LZ, DRTM doesn't
```
Something like this?

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 15]
```diff
@@ -0,0 +1,419 @@
...
+exist. Although it seems inconspicuous, it does fundamental job in secure boot
+process.
+
+#### Linux kernel
+
+Linux kernel is a core of all Linux based operating systems. Actually, it is
+operating system without any additional applications. It can be modified to
+given platform by including and excluding particular drivers, called modules. In
+general, it is done to limit size of kernel image and adjust its performance.
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Linux kernel also does some measurements in the Trenchboot Secure Launch process.

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 16]
```diff
@@ -0,0 +1,419 @@
...
+
+## Bring it all together
+
+I have mentioned and described all parts of system which you need to know, to
+understand TrenchBoot and DRTM. But how they are related and how to build such
+system? I said that package manager in NixOS is very important and we will
+commonly use it. And now I want to show you how.
+
+First, we have 'clean' NixOS operating system installed on hard drive. It has
+default Linux kernel and initrd. From GRUB we can boot to it and... that's it.
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
default Linux kernel and initrd. From GRUB we can boot it and... that's it.
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 17]
```diff
@@ -0,0 +1,419 @@
...
+No DRTM, no measurements and no secure. But, with custom package management, we
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
No DRTM, no measurements and no security. But, with custom package management, we
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 18]
```diff
@@ -0,0 +1,419 @@
...
+can **enable DRTM** and **prepare all components** directly from operating
+system. That's what we did. We have prepared, in separate packages, all elements
+mentioned above:
+
+- GRUB-trenchboot - it updates GRUB, so it boot `slaunch` and in consequence LZ
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
- GRUB-trenchboot - it updates GRUB, so it boots `slaunch` and in consequence LZ
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 19]
```diff
@@ -0,0 +1,419 @@
...
+- LZ-trenchboot - it adds LZ which measures and runs Linux kernel (enables DRTM)
+- linux-kernel-trenchboot - customized NixOS kernel
+- initrd-trenchboot - customized initrd
+
+Installing those packages will create a target system with enabled DRTM. How to
+do this in practice and verify if it really works will be discussed in next
+article! So definitely, wait for next post to use this knowledge in practice.
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: It is a little bit weird to encourage somebody to do something secure, but wait a few days to know what it will do to your machine... What should we do about it? @pietrushnic cc

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 20]
```diff
@@ -0,0 +1,419 @@
...
+```bash
+nix-env -f . -iA flashrom2
+installing 'flashrom2-2.0'
+building '/nix/store/hbxi8l93x2qv3kzg7kjpfaa6pmkij48f-user-environment.drv'...
+created 21 symlinks in user environment
+```
+
+You have just built your first package in NixOS. Moreover, it is your custom
+build! However, as you can see entire process is non-trivial. Essential part
+each pakcage is `stdenv` library, which
```
Comment (krystian-hebel) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
of each package is `stdenv` library, its
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 13 above)
Comment (krystian-hebel) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Please double check if I wrote it correctly
```suggestion
executes AMD's dedicated machine instruction `SKINIT`. Treat is 'DLEvent'
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 21]
```diff
@@ -0,0 +1,419 @@
...
+To better understand next section, first I will introduce operating system which
+we use. Our choice is [NixOS](https://nixos.org/nixos/) which is a Linux
+distribution based on unusual (and reliable) package management. You can install
+tools from official packages or build your own ones. Second feature will be
+strongly utilizes in entire project. Later, I will show you how to build custom
+package on `flashrom` example.
+
+> Of course, you can use any other Linux distribution. However, every procedure
+will be carried out by us in NixOS. If you won't have all dedicated tools or
+wrong configuration, we can't provide full reliability.
```
[the remark this answers is not in the document]
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: @krystian-hebel I think original message reflects what I meant. No dedicated tools or wrong configuration will cause non-reliability.

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 13 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Actually, I wanted to define SKINIT as 'pre-DRTM'. It is simplification, but should be suitable for non-technical users too.

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 1 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 2 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Logo is added

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 3 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 4 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 5 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 15 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: I don't mention it here, but later in 'linux-trenchboot' package description.

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 16 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 17 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 18 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 19 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: As both articles will be published in the same day, I rephrase it to indicate to next article immediately.

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 22]
```diff
@@ -0,0 +1,419 @@
...
+expanded. Eventually, there will be few different devices with different
+characteristic. For now, we use **PC Engines apu2** which has **AMD GX-412TC**
+processor. We use connection via serial port, so every relevant output will
+presented in form of logs.
+
+## Operating system
+
+To better understand next section, first I will introduce operating system which
+we use. Our choice is [NixOS](https://nixos.org/nixos/) which is a Linux
+distribution based on unusual (and reliable) package management. You can install
```
[the remark this answers is not in the document]
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 23]
```diff
@@ -0,0 +1,419 @@
...
+firmware and software components we are using. The most essential parts of
+system will be also described in details. It's important to understand their
+role, so you will be able to, fully consciously, reproduce the environment.
+
+## Testing platform
+
+As you already know, our solution is built for AMD processors. Therefore, for
+tests we use suitable platforms. Our hardware infrastructure is still being
+expanded. Eventually, there will be few different devices with different
+characteristic. For now, we use **PC Engines apu2** which has **AMD GX-412TC**
```
[the remark this answers is not in the document]
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 24]
```diff
@@ -0,0 +1,419 @@
...
+customize it in a runtime.
+
+>Important: NixOS installation is not interactive as in most Linux distributions
+(e.g. Debian). Therefore, pay attention to typed commands and configurations!
+
+1. Download minimal installer image from the [website](https://nixos.org/nixos/download.html).
+
+2. Flash USB drive with the image.
+
+3. Insert USB drive and choose proper boot partition.
```
[the remark this answers is not in the document]
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 25]
```diff
@@ -0,0 +1,419 @@
...
+nixos login: nixos (automatic login)
+```
+
+6. Create legacy boot partition
+
+  >Be careful and choose correct /dev/sdX device. In our case it is `sda`, which
+  is SSD disk.
+
+  - Create a MBR partition table
+  ```bash
```
[the remark this answers is not in the document]
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Fixed in each occurrence.

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 20 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 26]
```diff
@@ -0,0 +1,429 @@
...
+role, so you will be able to reproduce the environment smoothly and consciously.
+
+## Testing platform
+
+As you already know, our solution is built for AMD processors. Therefore, for
+tests we use suitable platforms. Our hardware infrastructure is still being
+expanded. Eventually, there will be few different devices with different
+characteristic. For now, we use **PC Engines apu2** which has **AMD GX-412TC**
+processor. We use connection via serial port, so every relevant output will
+presented in form of logs. What is important, our platform is equipped in
```
Comment (krystian-hebel) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
presented in form of logs. What is important, our platform is equipped with
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (krystian-hebel) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: @piotr-kleins the way I read it is `If you won't have (...) wrong configuration...`

[the remark this answers is not in the document]
Comment (krystian-hebel) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Also, `won't` -> `don't` would be better IMO

[the remark this answers is not in the document]
Comment (miczyg1) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: don't is better

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 8 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 9 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 10 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Done

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 12 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Give reference to PC Engines/coreboot repository

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md (hunk 13 above)
Comment (piotr-kleins) [reply blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]: Yes, I rephrased it

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 27]
```diff
@@ -0,0 +1,432 @@
...
+see the flow and relations between them.
+
+#### BIOS
+
+Our solution is open-source, so we also use such firmware (BIOS) - **coreboot**.
+Also, as we are maintainers of coreboot for PC Engines platforms, we use
+[PC Engines coreboot](https://github.com/pcengines/coreboot) and its mainline
+release. Every change appears there, so it is definitely place where you will
+find up-to-date firmware. I won't describe coreboot itself here. What you need
+to know, it is first code which is running during boot process. It initialize
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
to know, it is first code which is running during boot process. It initializes
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 28]
```diff
@@ -0,0 +1,432 @@
...
+exist. Although it seems inconspicuous, it does fundamental job in secure boot
+process.
+
+#### Linux kernel
+
+Linux kernel is a core of all Linux based operating systems. Actually, it is
+operating system without any additional applications. It can be modified to
+given platform by including and excluding particular drivers, called modules. In
+general, it is done to limit size of kernel image and adjust its performance. In
+Trenchboot Secure Launch process it perform additional operations - makes some
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
Trenchboot Secure Launch process it performs additional operations - makes some
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 29]
```diff
@@ -0,0 +1,432 @@
...
+measurements. But more about it, we will discuss later.
+
+#### initrd
+
+initrd stands for *initial ramdisk*. It is image used by bootloader to mount
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
initrd stands for *initial ramdisk*. It is an image used by bootloader to mount
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 30]
```diff
@@ -0,0 +1,432 @@
...
+understand TrenchBoot and DRTM. But how they are related and how to build such
+system? I said that package manager in NixOS is very important and we will
+commonly use it. And now I want to show you how.
+
+First, we have 'clean' NixOS operating system installed on hard drive. It has
+default Linux kernel and initrd. From GRUB we can boot it and... that's it. No
+DRTM, no measurements and no security. But, with custom package management, we
+can **enable DRTM** and **prepare all components** directly from operating
+system. That's what we did. We have prepared, in separate packages, all elements
+mentioned above:
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
mentioned below:
```

File: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md [hunk 31]
```diff
@@ -0,0 +1,432 @@
...
+
+- GRUB-trenchboot - it updates GRUB, so it boots `slaunch` and in consequence LZ
+- LZ-trenchboot - it adds LZ which measures and runs Linux kernel (enables DRTM)
+- linux-kernel-trenchboot - customized NixOS kernel which also make some
```
Comment (miczyg1) [inline blog/content/post/2020-03-27-trenchboot-nlnet-lz.md]:
```suggestion
- linux-kernel-trenchboot - customized NixOS kernel which also makes some
```
